### ISSUE TITLE
OpenBSD support

### ISSUE THREAD

**Issue by rdagnelie** [romkatv/gitstatus#84]
I would like to use powerlevel10k on OpenBSD, unfortunately it does not work out of the box:

```
    +gitstatus_start_impl:15> [[ -x /home/john/.oh-my-zsh/custom/themes/powerlevel10k/gitstatus/bin/gitstatusd-openbsd-amd64 ]]
                               ^ this command failed

```

Obviously, the binary is not builded for openbsd so the folder bin does not contain it.
When I try to build the binary with the build.zsh, I first meet a problem on the OS detection / CPU which i dirty patched it like this:

```
case $OS in
  FreeBSD|OpenBSD) CPUS=$(sysctl -n hw.ncpu);;
  *) CPUS=$(getconf _NPROCESSORS_ONLN);;
esac

```

But Im stucked right now with another error:

`*** Parse error in /tmp/gitstatus/gitstatus: Missing dependency operator (Makefile:27)
`

Do you think there is too much work to achieve the compilation process or can you help me to achieve openbsd compatibility ? (Im not a C/C++ developer)

**Comment by romkatv** [comment]
> But Im stucked right now with another error:

This error indicates that `Makefile` is being processed with `make` rather than `gmake`. If you look at `build.zsh`, you'll see that it mentions FreeBSD in two places. You've changed one of them but not the other.

I just submitted https://github.com/romkatv/gitstatus/commit/ccfd3d2f48a792f26f9c8e73014893d285487849 that replaces both `FreeBSD` instances in `build.zsh` with `*BSD`. I haven't tested it though. Let me know whether it works for you.

P.S.

It's unlikely that you'll need to know anything about C++ to successfully build gitstatusd. You might need to know a bit of zsh (to read `build.zsh`) and some general infrastructure such as GNU make, CMake, ld, etc.

**Comment by rdagnelie** [comment]
still KO with:

```
cc1plus: error: unrecognized command line option "-std=c++14"g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -std=c++14 -funsigned-char -O3 -DNDEBUG -Wall -Werror -fsized-deallocation  -MM -MT obj/git.o src/git.cc >obj/git.dep

cc1plus: error: unrecognized command line option "-fsized-deallocation"
gmake: *** [Makefile:21: obj/arena.o] Error 1
gmake: *** Waiting for unfinished jobs....
cc1plus: error: unrecognized command line option "-std=c++14"cc1plus: error: unrecognized command line option "-std=c++14"
cc1plus: error: unrecognized command line option "-fsized-deallocation"

cc1plus: error: unrecognized command line option "-fsized-deallocation"
gmake: *** [Makefile:21: obj/check_dir_mtime.o] Error 1
cc1plus: error: unrecognized command line option "-std=c++14"
cc1plus: error: unrecognized command line option "-fsized-deallocation"
gmake: *** [Makefile:21: obj/dir.o] Error 1
gmake: *** [Makefile:21: obj/git.o] Error 1

```

I also tried to remove options std and fsized but I get new different errors and i don't if it's the right way:

```
Building gitstatus in /tmp/gitstatus ...
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/arena.o src/arena.cc >obj/arena.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/check_dir_mtime.o src/check_dir_mtime.cc >obj/check_dir_mtime.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/dir.o src/dir.cc >obj/dir.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/git.o src/git.cc >obj/git.dep
In file included from src/arena.cc:18:
src/arena.h:27:25: error: type_traits.h: No such file or directory
In file included from src/dir.h:24,
                 from src/dir.cc:18:
src/arena.h:27:25: error: type_traits.h: No such file or directory
src/arena.cc:21:23: error: type_traits: No such file or directory
src/dir.cc:21:18: error: atomic: No such file or directory
gmake: *** [Makefile:21: obj/arena.o] Error 1
gmake: *** Waiting for unfinished jobs....
gmake: *** [Makefile:21: obj/dir.o] Error 1
In file included from src/dir.h:24,
                 from src/check_dir_mtime.cc:34:
src/arena.h:27:25: error: type_traits.h: No such file or directory
In file included from src/check_dir_mtime.cc:36:
src/print.h:26:23: error: type_traits: No such file or directory
In file included from src/git.cc:25:
src/print.h:26:23: error: type_traits: No such file or directory
gmake: *** [Makefile:21: obj/git.o] Error 1
gmake: *** [Makefile:21: obj/check_dir_mtime.o] Error 1

```


Seems the makefile need also some tweaks, keep digging....

**Comment by romkatv** [comment]
These errors say that your GCC is too old. What's the output of `g++ --version`?

**Comment by rdagnelie** [comment]
```

 john@doe  ~/git/gitstatus   master ●  g++ --version
g++ (GCC) 4.2.1 20070719
Copyright (C) 2007 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

 john@doe  ~/git/gitstatus   master ●  g++ --version
g++ (GCC) 4.2.1 20070719
Copyright (C) 2007 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by romkatv** [comment]
Yes, this GCC is ancient. You need to upgrade.

It's very unlikely that you need to change anything in the build script, Makefile or the C++ sources. Assume that all errors likely mean you are missing software or it's to old.

**Comment by rdagnelie** [comment]
I also tried with egcc from a comment i found:
```

# System compiler is gcc 4.2.1 up to OpenBSD 6.0.
# For proper C++11 support, we need to use a newer gcc from ports/packages,
# where compiler commands are renamed to egcc/eg++. Therefore, redefine
# mkspecs/common/g++-base.conf compiler commands

```

So i changed it in the Makefile but same error:

```
 john@doe  ~/git/gitstatus   master ●  /home/john/git/gitstatus/build.zsh
Building gitstatus in /tmp/gitstatus ...
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/arena.o src/arena.cc >obj/arena.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/check_dir_mtime.o src/check_dir_mtime.cc >obj/check_dir_mtime.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/dir.o src/dir.cc >obj/dir.dep
g++  -I/tmp/gitstatus/libgit2/include -DGITSTATUS_ZERO_NSEC -funsigned-char -O3 -DNDEBUG -Wall -Werror  -MM -MT obj/git.o src/git.cc >obj/git.dep
In file included from src/arena.cc:18:
src/arena.h:27:25: error: type_traits.h: No such file or directory
In file included from src/dir.h:24,
                 from src/dir.cc:18:
src/arena.h:27:25: error: type_traits.h: No such file or directory
src/arena.cc:21:23: error: type_traits: No such file or directory
gmake: *** [Makefile:21: obj/arena.o] Error 1
gmake: *** Waiting for unfinished jobs....
src/dir.cc:21:18: error: atomic: No such file or directory
gmake: *** [Makefile:21: obj/dir.o] Error 1
In file included from src/git.cc:25:
src/print.h:26:23: error: type_traits: No such file or directory
gmake: *** [Makefile:21: obj/git.o] Error 1
In file included from src/dir.h:24,
                 from src/check_dir_mtime.cc:34:
src/arena.h:27:25: error: type_traits.h: No such file or directory
In file included from src/check_dir_mtime.cc:36:
src/print.h:26:23: error: type_traits: No such file or directory
gmake: *** [Makefile:21: obj/check_dir_mtime.o] Error 1
 john@doe  ~/git/gitstatus   master ●  eg++ --version
eg++ (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```


My makefile:
```
APPNAME ?= gitstatusd

CXX ?= eg++

CXXFLAGS += -funsigned-char -O3 -DNDEBUG -Wall -Werror # -g -fsanitize=thread
LDFLAGS += -pthread # -fsanitize=thread
LDLIBS += -lgit2 # -lprofiler -lunwind

SRCS := $(shell find src -name "*.cc")
OBJS := $(patsubst src/%.cc, obj/%.o, $(SRCS))

all: $(APPNAME)

$(APPNAME): $(OBJS)
        $(CXX) $(OBJS) $(LDFLAGS) $(LDLIBS) -o $(APPNAME)

obj:
        mkdir -p obj

obj/%.o: src/%.cc Makefile | obj
        $(CXX) $(CXXFLAGS) -MM -MT $@ src/$*.cc >obj/$*.dep
        $(CXX) $(CXXFLAGS) -Wall -c -o $@ src/$*.cc

clean:
        rm -rf obj

-include $(OBJS:.o=.dep)

```

**Comment by romkatv** [comment]
Too bad. I propose to close this issue. I don't feel like installing openbsd myself and remote compiling doesn't seem like effective use of my and your time. Fair enough?

**Comment by rdagnelie** [comment]
ok I understand i will PR if i found a clean solution to provide openbsd support :)

**Comment by romkatv** [comment]
By the way, the error messages make no sense. Are you sure you haven't modified any C++ sources?

I repeat that is extremely unlikely that you need to modify anything. You can set your preferred compiler by exporting `CXX`.

**Comment by rdagnelie** [comment]
I succeeded:

built: /tmp/gitstatus/gitstatus/bin/gitstatusd-openbsd-amd64
self-check successful

:)

**Comment by romkatv** [comment]
> ok I understand i will PR if i found a clean solution to provide openbsd support :)

Thanks. I'll repeat once more just in case. There is very high chance that you don't need to modify anything in gitstatus in order to build it. Working from this assumption should set you on the right path.

**Comment by rdagnelie** [comment]
i dirty patched:

function build_gitstatus() {
  cd $DIR
  #[[ -n gitstatus(#qFN) ]] || git clone --depth 1 $GITSTATUS_REPO_URL
  cd gitstatus
  local cxx=${CXX:-'eg++'}

I replaced g++ by eg++, i did not really read your script, could we overload this part in the case loop ?

**Comment by romkatv** [comment]
> I succeeded:
> 
> built: /tmp/gitstatus/gitstatus/bin/gitstatusd-openbsd-amd64
> self-check successful
> 
> :)

👍 

I gather it builds out of the box without any changes?

**Comment by jdelkins** [comment]
I could not get the `build` script to work on OpenBSD 6.8 after numerous attempts. I would have liked to get it working in order to submit a PR, but I got tired of trying to figure out what it's doing and I gave up with it. I was, however, able to get a successful build with the following procedure. Documenting here in case it helps anyone else.

1. Install the needed build dependencies, most importantly `gmake` and `cmake` but there may be others I forgot about.

2. OpenBSD's `c++` doesn't work, so `pkg_add g++`, which installs the binary as `/usr/local/bin/eg++`
 
3. Extract the the maintainer's private fork of `libgit2` into `gitstatus/deps/libgit2`

4. `mkdir deps/libgit2/build && cd deps/libgit2/build`

5. Do the following. Note that `build` normally sets `-DTHREADSAFE=OFF` but that results in a linking error (`git___init`). `-DTHREADSAFE=ON` seems to work, however.

		cmake -DCMAKE_BUILD_TYPE=Release -DZERO_NSEC=ON -DTHREADSAFE=ON -DUSE_BUNDLED_ZLIB=ON -DREGEX_BACKEND=builtin -DUSE_HTTP_PARSER=builtin -DUSE_SSH=OFF -DBUILD_CLAR=OFF -DUSE_GSSAPI=OFF -DUSE_NTLMCLIENT=OFF -DBUILD_SHARED_LIBS=OFF ..

6. `cd src && gmake`

7. `cd ../../../..` (back to gitstatus directory), then build as follows

		CXXFLAGS=-Ideps/libgit2/include LDLIBS="-Ldeps/libgit2/build -lssl -lcrypto" CXX=eg++ gmake

This should produce a working `usrbin/gitstatusd`
